compiler/src/dotty/tools/dotc/typer/Inferencing.scala

```diff
@@ -154,15 +157,58 @@ object Inferencing {
    *  their lower bound. Record whether successful.
    *  2nd Phase: If first phase was successful, instantiate all remaining type variables
    *  to their upper bound.
+   *
+   *  Instance types can be improved by replacing covariant occurrences of Nothing
+   *  with fresh type variables, if `force` allows this in its `canImprove` implementation.
    */
   private class IsFullyDefinedAccumulator(force: ForceDegree.Value, minimizeSelected: Boolean = false)
     (using Context) extends TypeAccumulator[Boolean] {
 
-    private def instantiate(tvar: TypeVar, fromBelow: Boolean): Type = {
+    /** Replace toplevel-covariant occurrences (i.e. covariant without double flips)
+     *  of Nothing by fresh type variables.
+     *  For singleton types and references to module classes: try to
+     *  improve the widened type. For module classes, the widened type
+     *  is the intersection of all its non-transparent parent types.
+     */
+    private def improve(tvar: TypeVar) = new TypeMap:
+      def apply(t: Type) = trace(i"improve $t", show = true):
+        def tryWidened(widened: Type): Type =
+          val improved = apply(widened)
+          if improved ne widened then improved else mapOver(t)
+        if variance > 0 then
+          t match
+            case t: TypeRef =>
+              if t.symbol == defn.NothingClass then
+                newTypeVar(TypeBounds.empty, nestingLevel = tvar.nestingLevel)
```

Comment:
Leaving the upper-bound as `Any` seeems too lax, in particular if the type parameter is higher-kinded this means we can infer something of the wrong kind, which at the very least leads to broken error messages:
```scala
class Foo[+T[_]]:
  def combine[T1[x] >: T[x]](x: T1[Int]): Foo[T1] = new Foo
object Foo:
  def empty: Foo[Nothing] = new Foo

object X:
  def test(xs: List[List[Int]]): Unit =
    xs.foldLeft(Foo.empty)((ys, x) =>
      ys.combine(x)
    )
```
```scala
-- [E007] Type Mismatch Error: tests/pos/folds.scala:31:16 ---------------------
31 |      ys.combine(List(x))
   |      ^^^^^^^^^^^^^^^^^^^
   |      Found:    Foo[List]
   |      Required: Foo[List[X0]]
```

Reply:
OK, maybe we can restrict this to non-hk types.